facilitator/src/gcp_oauth.rs

```diff
@@ -84,17 +91,231 @@ struct ServiceAccountKeyFile {
+    /// ureq::Response, whose body will be an OauthTokenResponse if the HTTP
+    /// call was successful.
+    fn account_token_from_gke_metadata_service(
+        &self,
+        agent: &RetryingAgent,
+        logger: &Logger,
+    ) -> Result<Response> {
+        debug!(
+            logger,
+            "obtaining default account token from GKE metadata service"
+        );
+
+        let mut request = agent.prepare_request(RequestParameters {
+            url: default_oauth_token_url(),
+            method: Method::Get,
+            ..Default::default()
+        })?;
+
+        request = request.set("Metadata-Flavor", "Google");
+
+        agent
+            .call(logger, &request)
+            .context("failed to query GKE metadata service")
+    }
+
+    /// Fetches the default account token from Google OAuth API using a JWT
+    /// constructed from the parameters in the provided key file. If the JWT is
+    /// successfully constructed, returns the ureq::Response whose body will be
+    /// an OauthTokenResponse if the HTTP call was successful, but may be an
+    /// error.
+    fn account_token_with_key_file(
+        &self,
+        key_file: &ServiceAccountKeyFile,
+        scope: &str,
+        agent: &RetryingAgent,
+        logger: &Logger,
+    ) -> Result<Response> {
+        debug!(logger, "obtaining account token from key file");
+        // We construct the JWT per Google documentation:
+        // https://developers.google.com/identity/protocols/oauth2/service-account#authorizingrequests
+        let mut header = Header::new(Algorithm::RS256);
+        header.kid = Some(key_file.private_key_id.to_owned());
+
+        // The iat and exp fields in a JWT are in seconds since UNIX epoch.
+        let now = Utc::now().timestamp();
+        let claims = Claims {
+            iss: key_file.client_email.to_owned(),
+            scope: scope.to_owned(),
+            aud: key_file.token_uri.to_owned(),
+            iat: now,
+            exp: now + 3600, // token expires in one hour
+        };
+
+        let encoding_key = EncodingKey::from_rsa_pem(key_file.private_key.as_bytes())
+            .context("failed to parse PEM RSA key")?;
+
+        let token =
+            encode(&header, &claims, &encoding_key).context("failed to construct and sign JWT")?;
+
+        let request_body = format!(
+            "grant_type={}&assertion={}",
+            urlencoding::encode("urn:ietf:params:oauth:grant-type:jwt-bearer"),
+            token
+        );
+
+        let request = agent.prepare_request(RequestParameters {
+            url: Url::parse(&key_file.token_uri).context(format!(
+                "failed to parse key_file.token_uri: {}",
+                &key_file.token_uri
+            ))?,
+            method: Method::Post,
+            ..Default::default()
+        })?;
+
+        let request = request.set("Content-Type", "application/x-www-form-urlencoded");
```

Comment:
You might be interested in https://docs.rs/ureq/2.1.1/ureq/struct.Request.html#method.send_form (note: you have to click open the `impl Request` because of some horrible rustdoc bug that I probably introduced)

Reply:
Good tip, this would save me the trouble of constructing the form request body myself.